syntax = "proto3";

package library;

// Timestamp message for datetime fields
message Timestamp {
  int64 seconds = 1;
  int32 nanos = 2;
}

// Member message
message Member {
  string id = 1;
  string first_name = 2;
  string last_name = 3;
  string full_name = 4;
  string email = 5;
  string phone = 6;
  string address = 7;
  string membership_number = 8;
  string membership_status = 9; // active, suspended, inactive
  string join_date = 10; // ISO 8601 date string
  string created_at = 11; // ISO 8601 datetime string
  string updated_at = 12; // ISO 8601 datetime string
  int32 active_borrowings_count = 13;
  int32 overdue_borrowings_count = 14;
}

// Member list response
message MemberList {
  repeated Member results = 1;
  int32 count = 2;
  string next = 3;
  string previous = 4;
}

// Book message
message Book {
  string id = 1;
  string isbn = 2;
  string title = 3;
  string author = 4;
  string publisher = 5;
  int32 publication_year = 6;
  string description = 7;
  int32 total_copies = 8;
  int32 available_copies = 9;
  string condition = 10; // excellent, good, fair, poor
  string language = 11;
  bool is_available = 12;
  int32 borrowing_count = 13;
  int32 active_borrowings_count = 14;
  string created_at = 15;
  string updated_at = 16;
}

// Book list response
message BookList {
  repeated Book results = 1;
  int32 count = 2;
  string next = 3;
  string previous = 4;
}

// Borrowing message
message Borrowing {
  string id = 1;
  string member = 2;
  string member_id = 3;
  string member_name = 4;
  string book = 5;
  string book_id = 6;
  string book_title = 7;
  string borrowed_at = 8; // ISO 8601 datetime string
  string due_date = 9; // ISO 8601 date string
  string returned_at = 10; // ISO 8601 datetime string (optional)
  string status = 11; // active, overdue, returned
  int32 days_until_due = 12;
  int32 days_overdue = 13;
  bool is_overdue = 14;
  string notes = 15;
  string created_at = 16;
  string updated_at = 17;
}

// Borrowing list response
message BorrowingList {
  repeated Borrowing results = 1;
  int32 count = 2;
  string next = 3;
  string previous = 4;
}

// Fine message
message Fine {
  string id = 1;
  string borrowing_id = 2;
  string member_id = 3;
  string member_name = 4;
  string book_id = 5;
  string book_title = 6;
  string amount = 7; // decimal as string
  string reason = 8;
  bool is_paid = 9;
  string paid_at = 10; // ISO 8601 datetime string (optional)
  string created_at = 11;
  string updated_at = 12;
}

// Fine list response
message FineList {
  repeated Fine results = 1;
  int32 count = 2;
  string next = 3;
  string previous = 4;
}

// Authentication messages
message LoginRequest {
  string email = 1;
  string password = 2;
}

message LoginResponse {
  string token = 1;
  Member member = 2;
  string message = 3;
}

message SignupRequest {
  string first_name = 1;
  string last_name = 2;
  string email = 3;
  string password = 4;
  string phone = 5;
  string address = 6;
}

message SignupResponse {
  string token = 1;
  Member member = 2;
  string message = 3;
}

// Error response
message ErrorResponse {
  string error = 1;
  string detail = 2;
  map<string, string> field_errors = 3;
}

// Generic message for actions
message ActionResponse {
  string message = 1;
  bool success = 2;
}

// Statistics message
message MemberStats {
  int32 total_borrowings = 1;
  int32 active_borrowings = 2;
  int32 overdue_borrowings = 3;
  int32 total_fines = 4;
  string total_fine_amount = 5;
}

message BookStats {
  int32 total_borrowings = 1;
  int32 active_borrowings = 2;
  double average_borrowing_duration = 3;
}
